### PR TITLE
Refactoring and cleanup of `species.param`

### PR DESCRIPTION
**Pull request by steindev**
Actions performed:
* Rename particle shapes to be consistent in naming -> requires renaming of the respective `*.hpp` files and classes therein
* Apply these changes to examples and test
* Add `getStringProperty()` to `FieldToParticleInterpolationNative` class to make it compilable, i.e. usable
* Clean examples from `species.param` files that were just old standards and never updated -> this actually means that most examples use `particles::shapes::TSC` instead of `particles::shapes::CIC` now

### CODE REVIEW

File: include/picongpu/param/species.param [hunk 1]
```diff
@@ -38,12 +38,23 @@ namespace picongpu
...
+ *  PIConGPU only implements the assignment functions.
+ *  **NOTE, the higher order shape names are redefined with release 0.6.0 in order to provide a consistent naming:
+ *      -> PQS is the name of the 3rd order shape (instead of PCS)
+ *      -> PCS is the name of the 4th order shape (instead of P4S)
+ *      -> P4S does not exist anymore
+ *  **
+ *  - particles::shapes::CIC : 1st order, assignment function is a piecewise linear spline
+ *  - particles::shapes::TSC : 2nd order, assignment function is a piecewise quadratic spline
+ *  - particles::shapes::PQS : 3rd order, assignment function is a piecewise cubic spline
+ *  - particles::shapes::PCS : 4th order, assignment function is a piecewise quartic spline
```
Comment (sbastrakov) [inline include/picongpu/param/species.param:57]: What does the `order` part of the comment add? The information on the assignment function is given after the comma. While out of context this `order` may be misunderstood as the cloud density order.

File: share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param [hunk 2]
```diff
@@ -17,54 +17,74 @@
...
- *  - particles::shapes::TSC : 2nd order
- *  - particles::shapes::PCS : 3rd order
- *  - particles::shapes::P4S : 4th order
- *
- *  example: using UsedParticleShape = particles::shapes::CIC;
+/** Particle Shape definitions
+ *  Particle shape naming is according to
+ *    Hockney R, Eastwood JW. Computer simulation using particles. IOP Pubslishing Ltd (1988). p. 144.
+ *  That is, the name of a particle shape is defined by the visual form of its cloud shape S,
+ *  which describes a particle's charge density distribution.
```
Comment (sbastrakov) [inline share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:44]: Macroparticle's ?

File: share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param [hunk 3]
```diff
@@ -17,54 +17,74 @@
...
+ *  However, the charge assigned to a mesh point due to a distribution of overlapping finite-sized
+ *  particles is obtained by evaluating and summing the assignment function W for all particles.
+ *  The assignment function is obtained by convolving the cloud shape with the top-hat function.
+ *  PIConGPU only implements the assignment functions.
```
Comment (sbastrakov) [inline share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:48]: I am not sure this information is relevant at this point. I would imagine a user cares about how to set the desired properties, not how it's implemented internally

File: share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param [hunk 4]
```diff
@@ -17,54 +17,74 @@
...
+ * - currentSolver::EsirkepovNative< SHAPE, STRATEGY > : generic version of currentSolverEsirkepov
  *   without optimization (~4x slower and needs more shared memory)
  *
  * STRATEGY (optional):
- * - currentSolver::strategy::StridedCachedSupercells
- * - currentSolver::strategy::CachedSupercells
- * - currentSolver::strategy::NonCachedSupercells
+ * - currentSolver::StridedCachedSupercells
+ * - currentSolver::CachedSupercells
+ * - currentSolver::NonCachedSupercells
```
Comment (sbastrakov) [inline share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:83]: Not sure if this is correct. That `strategy` namespace seems still in place

**Comment by sbastrakov** [comment]
@steindev thanks, I like these changes in principle, some small notes above. However, why do they all have to be in the same PR? To me it seems like the changes are of 3 kinds:

- Renaming of shapes and updating comments on that
- Other cleaup in species.param
- Changes to `FieldToParticleInterpolationNative `

Couldn't they all be separate PRs ?

File: share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param (hunk 3 above)
Comment (psychocoderHPC) [reply share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:48]: I would also move such information to the documentation instead of having it here.
Should we also shorten this section and move it to the documentation and link the documentation here:

```
...

**WARNING** the shape naming has changed and is not consistent too PIConGPU versions before 0.6.0. see readthedocslink

  - particles::shapes::CIC : 1st order, assignment function is a piecewise linear spline
  - particles::shapes::TSC : 2nd order, assignment function is a piecewise quadratic spline
  - particles::shapes::PQS : 3rd order, assignment function is a piecewise cubic spline
  - particles::shapes::PCS : 4th order, assignment function is a piecewise quartic spline
```

File: include/picongpu/param/species.param (hunk 1 above)
Comment (steindev) [reply include/picongpu/param/species.param:57]: Well, nothing anymore. Just didn't remove it. Will do now.

File: share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param (hunk 3 above)
Comment (steindev) [reply share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:48]: I moved most of the text to the general file documentation, where I think it belongs and is visible at readthedocs. I also think we can leave the explaining sentence on PIConGPU imlementing only assignment functions at this general file documentation level, as it will help users/developers understanding what PIConGPU does. I agree, it is superfluous as a comment to `using UsedParticleShape`.

File: share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param (hunk 4 above)
Comment (steindev) [reply share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:83]: Hmm that happened accidentally when replacing the old file in the test by the new one. As this usage is explained in the standard `species.param`, I assume it is the recommended way.
And that the documentation in this test's `species.param` has never been updated but reflects an old standard.

**Comment by steindev** [comment]
@sbastrakov and @psychocoderHPC: I updated the documentation in `species.param` according to your comments. Please have a look. If you think it is ok now, I will also update all the other `species.param` files in the examples and push again.

**Comment by steindev** [comment]
> However, why do they all have to be in the same PR? [...]
> Couldn't they all be separate PRs ?

Well I grouped in 4 different commits, so they can end up in PIConGPU as 4 different commits, which will be the same result as using individual PRs. In the end, it will therefore make no difference. But as all these logically belong together and are in itself just small changes, I thought it would be sufficient to have only on PR for these and thereby reduce the additional effort.

File: share/picongpu/tests/XrayScattering/include/picongpu/param/species.param [hunk 5]
```diff
@@ -1,4 +1,4 @@
-/* Copyright 2014-2020 Rene Widera, Richard Pausch, Pawel Ordyna
```
Comment (PrometheusPi) [inline share/picongpu/tests/XrayScattering/include/picongpu/param/species.param:1]: why is Annegret listed here as co-author 
(applies to various other files as well)

**Review by PrometheusPi — APPROVED** [review]
Looks good to me, just a question regarding multiple new authors in various files.

File: share/picongpu/tests/XrayScattering/include/picongpu/param/species.param (hunk 5 above)
Comment (steindev) [reply share/picongpu/tests/XrayScattering/include/picongpu/param/species.param:1]: She implemented the Higuera-Cary particle pusher which can be selected in the `species.param`. Since she edited this file to make it available she is an author. And since I copied the new standard file to all examples, she is now an author in several files were she hasn't been an author before.

File: include/picongpu/param/species.param [hunk 6]
```diff
@@ -19,9 +19,25 @@
...
+ * However, the charge assigned to a mesh point due to a distribution of overlapping finite-sized
+ * macroparticles is obtained by evaluating and summing the assignment function W for all particles.
+ * The assignment function is obtained by convolving the cloud shape with the top-hat function.
+ * PIConGPU only implements the assignment functions.
+ *
+ * @warning
+ * The higher order shape names are redefined with release 0.6.0 in order to provide a consistent naming:
+ *     -> PQS is the name of the 3rd order shape (instead of PCS)
+ *     -> PCS is the name of the 4th order shape (instead of P4S)
+ *     -> P4S does not exist anymore
```
Comment (sbastrakov) [inline include/picongpu/param/species.param:40]: I think this piece is not self-consistent. From the first sentence a reader may assume everything is about cloud shapes. But these `3rd` and `4th order shape` are **not** shape orders. Rather, they are assignment function orders

**Review by sbastrakov — CHANGES_REQUESTED** [review]

File: include/picongpu/param/species.param (hunk 6 above)
Comment (sbastrakov) [reply include/picongpu/param/species.param:40]: I think the text should be corrected to have `3rd order assignment function` instead of `3rd order shape`

**Comment by steindev** [comment]
@sbastrakov, @psychocoderHPC: I expanded the documentation and again updated `species.param` accordingly. As before, I still need to update all other `species.param`s which I will do after your review.

**Review by sbastrakov — DISMISSED** [review]

File: share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param (hunk 4 above)
Comment (psychocoderHPC) [reply share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:83]: `currentSolver::strategy::` must be included!

File: share/picongpu/examples/SingleParticleTest/include/picongpu/param/species.param [hunk 7]
```diff
@@ -17,51 +17,71 @@
...
  * For development purposes:
- * - currentSolver::currentSolver::EsirkepovNative< SHAPE, STRATEGY > : generic version of currentSolverEsirkepov
+ * - currentSolver::EsirkepovNative< SHAPE, STRATEGY > : generic version of currentSolverEsirkepov
  *   without optimization (~4x slower and needs more shared memory)
  *
  * STRATEGY (optional):
- * - currentSolver::strategy::StridedCachedSupercells
- * - currentSolver::strategy::CachedSupercells
- * - currentSolver::strategy::NonCachedSupercells
+ * - currentSolver::StridedCachedSupercells
```
Comment (psychocoderHPC) [inline share/picongpu/examples/SingleParticleTest/include/picongpu/param/species.param:78]: Please revert this change it must be `currentSolver::strategy::StridedCachedSupercells`

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

File: share/picongpu/examples/SingleParticleTest/include/picongpu/param/species.param (hunk 7 above)
Comment (sbastrakov) [reply share/picongpu/examples/SingleParticleTest/include/picongpu/param/species.param:78]: Wait, I was already asking about it for the very first version of this PR.

File: share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param (hunk 4 above)
Comment (steindev) [reply share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:83]: :ok:

**Comment by steindev** [comment]
@psychocoderHPC I addressed your comments and included the refactored `species.param` into all examples and tests. CIs are happy, too. Ready to merge?!

File: docs/source/models/pic.rst [hunk 8]
```diff
@@ -64,20 +64,12 @@ Electro-Magnetic PIC Method
...
-
-   S^0(x) = \big\{ \substack{1 \qquad \text{if}~0 \le x \lt 1\\ 0 \qquad \text{else}}
-
-   S^n(x) = \left(S^{n-1} * S^0\right)(x) = \int_{x-1}^x S^{n-1}(\xi) d\xi
-
-PIConGPU implements these up to order :math:`n=4`.
-The three dimensional marker shape is a multiplicative union of B-splines :math:`S^n(x,y,z) = S^n(x) S^n(y) S^n(z)`.
+A marker has a finite-size and a velocity, such that it can be regarded as a cloud of particles, whose center of mass is the marker's position and whose mean velocity is the marker's velocity.
+The cloud shape :math:`S^n(x)` of order :math:`n` of a marker describes its charge density distribution.
+See :ref:`section Hierarchy of charge assignment schemes <model-shapes>` for a list of available marker shapes in PIConGPU.
```
Comment (PrometheusPi) [inline docs/source/models/pic.rst:72]: typo: `Hierarchy` -> `hierarchy`

File: docs/source/models/shapes.rst [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+In PIConGPU, the cloud shapes :math:`S^n(x)` are pre-integrated to *assignment functions* :math:`W^n(x)`.
+
+.. math::
+   W^n(x) = \Pi(x) \ast S^n(x) = \int\limits_{-\infty}^{+\infty} \Pi(x^\prime) S^n(x^\prime - x) dx^\prime\,, \text{ where }
+   \Pi(x) = \left\{\begin{array}{ll}
+            0 & |x| \gt \frac{1}{2} \\
+            \frac{1}{2} & |x| = \frac{1}{2} \\
+            1 & |x| \lt \frac{1}{2}
+      \end{array}\right.\,.
```
Comment (PrometheusPi) [inline docs/source/models/shapes.rst:16]: Why do you end the equation with a dot? 
It is not the end of a sentence.

File: docs/source/models/shapes.rst [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+Evaluating the assignment functions at mesh points directly provides the fraction of charge from the marker assigned to that point.
+
+The assignment functions are implemented as B-splines.
+The zeroth order assignment function :math:`W^0` is the top-hat function :math:`\Pi`.
+It represents charge assignment to the nearest mesh point only, resulting in a stepwise charge density distribution.
+Therefore, it should not be used.
+The assignment function of order :math:`n` is generated by convolution of the assignment function of order :math:`n-1` with the top-hat function
+
+.. math::
+
```
Comment (PrometheusPi) [inline docs/source/models/shapes.rst:29]: does this space breaks the math mode?

File: share/picongpu/tests/XrayScattering/include/picongpu/param/species.param [hunk 11]
```diff
@@ -1,4 +1,4 @@
...
+/* Copyright 2014-2020 Rene Widera, Richard Pausch, Annegret Roeszler, Klaus Steiniger
```
Comment (PrometheusPi) [inline share/picongpu/tests/XrayScattering/include/picongpu/param/species.param:1]: You accidentally removed @pordyna

File: share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param [hunk 12]
```diff
@@ -17,54 +17,74 @@
...
+ * - currentSolver::EsirkepovNative< SHAPE, STRATEGY > : generic version of currentSolverEsirkepov
  *   without optimization (~4x slower and needs more shared memory)
  *
  * STRATEGY (optional):
  * - currentSolver::strategy::StridedCachedSupercells
  * - currentSolver::strategy::CachedSupercells
  * - currentSolver::strategy::NonCachedSupercells
  */
-using UsedParticleCurrentSolver = currentSolver::EmZ<UsedParticleShape>;
+using UsedParticleCurrentSolver = currentSolver::Esirkepov< UsedParticleShape >;
```
Comment (PrometheusPi) [inline share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:85]: Why not EmZ anymore?

File: share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param [hunk 13]
```diff
@@ -1,4 +1,4 @@
-/* Copyright 2014-2020 Rene Widera, Richard Pausch
+/* Copyright 2014-2020 Rene Widera, Richard Pausch, Annegret Roeszler, Klaus Steiniger
```
Comment (steindev) [inline share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:1]: Add sbastrakov

File: docs/source/models/pic.rst (hunk 8 above)
Comment (steindev) [reply docs/source/models/pic.rst:72]: I will correct to the correct title with capital nouns.

File: docs/source/models/shapes.rst (hunk 10 above)
Comment (steindev) [reply docs/source/models/shapes.rst:29]: No.

File: share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param (hunk 12 above)
Comment (steindev) [reply share/picongpu/tests/compileParticlePusher/include/picongpu/param/species.param:85]: My fault...

**Review by PrometheusPi — APPROVED** [review]

**Comment by psychocoderHPC** [comment]
grrrr :angry: Searched over 2h why my updated HIP branch is now slower than the old code base. After a diff I realiced tha twe change the shape names and I simulated with the shape `PCS` which has now a new order.
@jkelling with the next SPEC benchmark setup update we need to change https://github.com/psychocoderHPC/picongpu/blob/f22742b9fbfad7a5d725b41727991f1f63c01b00/share/picongpu/examples/SPECBenchmark/include/picongpu/param/species.param#L48 to `PQS`

**Comment by steindev** [comment]
:rofl: Did you use the TWEAC setup for testing, not any standard example? Wouldn't have expected that any test setup uses the higher order shapes...

But 2h is a significant loss these days :grimacing:

**Comment by psychocoderHPC** [comment]
> 🤣 Did you use the TWEAC setup for testing, not any standard example? Wouldn't have expected that any test setup uses the higher order shapes...
> 
> But 2h is a significant loss these days 😬

Yes the SPEC benchmark is currently not part of the examples. The advantage is that it requires only single precision and the amount of memory is scaling linear with the volume per GPU. There is one update required before I can add this example to our codebase.
